### PR TITLE
refactor(popper): Update to use react portal as default

### PR DESCRIPTION
**Pull request by mannycarrera4**
## Summary

Fixes: https://github.com/Workday/canvas-kit/issues/517

## Checklist

- [ ] branch has been rebased on the latest master commit
- [ ] tests are changed or added
- [ ] `yarn test` passes
- [ ] all (dev)dependencies that the module needs is added to its `package.json`
- [ ] code has been documented and, if applicable, usage described in README.md
- [ ] module has been added to `canvas-kit-react` and/or `canvas-kit-css` universal modules, if
      applicable
- [ ] design approved final implementation
- [ ] a11y approved final implementation
- [ ] code adheres to the [API & Pattern guidelines](../API_PATTERN_GUIDELINES.md)

### CODE REVIEW

File: modules/common/react/lib/Popper.tsx [hunk 1]
```diff
@@ -15,9 +15,16 @@ export interface PopperProps extends React.HTMLAttributes<HTMLDivElement> {
...
+   * The containing element for Popper elements. The Popper uses
+   * {@link https://reactjs.org/docs/portals.html Portals} to place the DOM elements
+   * of the Popper in a different place in the DOM to prevent issues with overflowed containers.
+   * When the popper is opened, `aria-hidden` will be added to siblings to hide background
+   * content from assistive technology like it is visibly hidden from sighted users. This property
+   * should be set to the element that the application root goes - not containing element of content.
+   * This should be a sibling or higher than the header and navigation elements of the application.
+   * @default document.body
    */
-  containerElement?: Element;
```
Comment (mannycarrera4) [inline modules/common/react/lib/Popper.tsx:20]: probably keep this as element?

**Comment by anicholls** [comment]
@mannycarrera4 Can you make sure you target `prerelease/v4`? Some of these updates (i.e. Popper) have already been made there

File: modules/common/react/README.md [hunk 2]
```diff
@@ -74,7 +74,13 @@ includes custom `data-*` attributes such as `data-test-id` to help facilitate au
 
 #### `containerElement: Element`
 
-> The element that contains the portal children when `portal` is true.
+> The containing element for Popper elements. The Popper uses {@link
```
Comment (anicholls) [inline modules/common/react/README.md:77]: Not valid link formatting for md

File: modules/common/react/README.md [hunk 3]
```diff
@@ -74,7 +74,13 @@ includes custom `data-*` attributes such as `data-test-id` to help facilitate au
...
+> https://reactjs.org/docs/portals.html Portals} to place the DOM elements of the Popper in a
+> different place in the DOM to prevent issues with overflowed containers. When the popper is
```
Comment (anicholls) [inline modules/common/react/README.md:79]: Nit: "in a different place in the dom" > "As the last child of the specified container element"

File: modules/common/react/README.md [hunk 4]
```diff
@@ -74,7 +74,13 @@ includes custom `data-*` attributes such as `data-test-id` to help facilitate au
...
+> opened, `aria-hidden` will be added to siblings to hide background content from assistive
```
Comment (anicholls) [inline modules/common/react/README.md:80]: I know we do this manually in modal - have we confirmed that it popper does this for us? Are we sure this is what we want for all popup use cases? If a tooltip pops up, I wouldn't call all of the rest of the content on the page hidden

File: modules/common/react/lib/Popper.tsx [hunk 5]
```diff
@@ -15,9 +15,16 @@ export interface PopperProps extends React.HTMLAttributes<HTMLDivElement> {
    */
   children: React.ReactNode;
   /**
-   * The element that contains the portal children when `portal` is true.
+   * The containing element for Popper elements. The Popper uses
+   * {@link https://reactjs.org/docs/portals.html Portals} to place the DOM elements
+   * of the Popper in a different place in the DOM to prevent issues with overflowed containers.
```
Comment (anicholls) [inline modules/common/react/lib/Popper.tsx:20]: Note to make the same copy updates here

File: modules/common/react/lib/Popper.tsx [hunk 6]
```diff
@@ -60,10 +61,6 @@ export class Popper extends React.PureComponent<PopperProps> {
       return null;
     }
 
-    if (!this.props.portal) {
-      return this.renderPopper();
-    }
-
     // do not use defaultProps for containerElement because document may not be statically available
     // at require time in some testing environments; instead we safely default at runtime
     return ReactDOM.createPortal(this.renderPopper(), this.props.containerElement || document.body);
```
Comment (anicholls) [inline modules/common/react/lib/Popper.tsx:66]: Can we combine `render` and `renderPopper` now? The only reason they were separated was due to the portal conditional

Comment (anicholls) [reply modules/common/react/lib/Popper.tsx:66]: Nit: Can we change `document.body` to a default prop?

File: modules/common/react/lib/Popper.tsx (hunk 1 above)
Comment (anicholls) [reply modules/common/react/lib/Popper.tsx:20]: `export function createPortal(children: ReactNode, container: Element, key?: null | string): ReactPortal;`
^ Types from React.createPortal, so yeah - let's keep as `Element`

**Comment by NicholasBoll** [comment]
I read the PR description and associated issue. We are already defaulting to using portals. This PR removes the option to not use a Portal. Is that what we wanted? There may be cases where not portalling content is desired.

**Comment by mannycarrera4** [comment]
@NicholasBoll from what I understood, we were just using portals, but this brings up the question of if we should keep the portaling option if they don't want to portal?

**Comment by NicholasBoll** [comment]
> @NicholasBoll from what I understood, we were just using portals, but this brings up the question of if we should keep the portaling option if they don't want to portal?

I'm not sure what that means. By default we use portals with the option to disable if necessary (in certain cases this can make focus management _much_ easier).

This API change still allows the original functionality of disabling the separation of child content from the parent (what portalling essentially does) by allowing a developer to set the `containerElement` to be the parent element, but this is much harder to do now:

```tsx
// before
<Popper portal={false}>

// after
<Portal containerElement={targetRef.current?.parentElement}
```

Full example here:
```tsx
// before
const BeforeComponent = () => {
  const targetRef = React.useRef<HTMLDivElement>(null)
  const [isOpen, setOpen] = React.useState(false)

  return (
    <>
      <div ref={targetRef} onClick={() => setOpen(true)}>Open</div>
      <Popper open={isOpen} anchorElement={targetRef.current} portal={false}>
        Popper content
      </Popper>
    </>
  )
}

// after
const AfterComponent = () => {
  const targetRef = React.useRef<HTMLDivElement>(null)
  const [isOpen, setOpen] = React.useState(false)

  return (
    <>
      <div ref={targetRef} onClick={() => setOpen(true)}>Open</div>
      <Popper open={isOpen} anchorElement={targetRef.current} containerElement={targetRef.current?.parentElement}>
        Popper content
      </Popper>
    </>
  )
}
```

The `after` isn't 100% accurate because we can no longer control the position in the DOM if we don't want to portal to a separate container. Portals will always be a the last child of a container.

**Comment by NicholasBoll** [comment]
@anicholls Is this an okay thing? No way to disable the use of Portal on a `Popper` component? There are valid reasons to not use it, but still have the niceties of using a position library.

**Comment by mannycarrera4** [comment]
@NicholasBoll I might have misinterpreted the issue, i think after your explanation on the example, i think we should just default to using portal instead of ripping it out

**Comment by mannycarrera4** [comment]
@anicholls @NicholasBoll I feel like we should just close this PR, most of the changes where in popper and they're being addressed in https://github.com/Workday/canvas-kit/pull/528 thoughts? I only really just updated some readmes

**Comment by anicholls** [comment]
@mannycarrera4 I'm fine with closing this in favor of #528. 

@NicholasBoll Can you make sure #517 gets closed as part of #528 as well? Squashing 10 tickets with one PR 😂

**Pull request closed without merging**